### DISCUSSION TITLE
Migrating from AMS

### DISCUSSION THREAD

**Discussion by dgmora** [ElMassimo/oj_serializers#11] [Q&A] [answered]
Hi, I've seen the guide and I'm trying to follow it but there are some points that I didn't find.

The are some things that I'm not sure how to change that are blocking me at the moment. One is that I thought that I could keep associations using AMS, but it seems that doesn't work:

```ruby
class MySerializer < Oj::Serializer
  belongs_to :user, serializer: MyAMSUserSerializer
end
```

This throws:

```
      NoMethodError:
        undefined method `one_as_hash' for MyAMSUserSerializer:Class
      # (eval):7:in `render_as_hash'
      # /Users/david/.local/share/rtx/installs/ruby/3.0.5/lib/ruby/gems/3.0.0/gems/oj_serializers-2.0.2/lib/oj_serializers/serializer.rb:209:in `one_as_hash'
      # /Users/david/.local/share/rtx/installs/ruby/3.0.5/lib/ruby/gems/3.0.0/gems/oj_serializers-2.0.2/lib/oj_serializers/serializer.rb:164:in `render'


```

I have `compat` required.

The other is that [`scope`](https://github.com/rails-api/active_model_serializers/blob/v0.10.4/docs/general/serializers.md#scope) isn't part of the DSL.

Some other points that I noticed that I had to change, that I think at some point I can manage to fix:
- `type` doesn't exist as part of the DSL
- - Haven't checked yet how to replace it yet, but `link` also isn't part of the DSL
- `unless` doesn't exist as opposite as `if`
- Had to replace `key:` with `as:`
 
Don't take me wrong, the library looks pretty promising and I understand replicating the whole AMS API isn't the goal, but I'm trying to figure out how to make the change!

**Comment by ElMassimo** [comment, marked as answer]
Hi David!

> undefined method `one_as_hash'

Thanks for reporting, this is a bug in the latest versions, have a fix here:

- https://github.com/ElMassimo/oj_serializers/pull/12

> `unless` doesn't exist as opposite as `if`
> Had to replace key: with as:

These were low-hanging fruit, implemented in:

- [feat: add support for key and unless in ams compat mode](https://github.com/ElMassimo/oj_serializers/pull/12/commits/b5477fe4ef9d85c5d03e780f50c41c8940c2b195)

---

`scope` is not implemented in compat-mode because it makes serializers context-dependent, and is too implicit, making it more difficult to understand and unit test them.

Usually, the best approach is to pass any additional information the serializer needs as options. As a temporary measure during migration, you could also use something like [`RequestLocals`](https://github.com/ElMassimo/request_store_rails) to share the current controller (or current user) and define helper methods in your base serializer.

---

> `type` and `link` don't exist as part of the DSL

This library targets `active_model_serializers `0.8-ish, not `0.10`.

**Reply by dgmora** [reply]
That was quick! I'll give it a try with these changes
